packages/graph-explorer/src/core/StateProvider/localDb.ts

```diff
@@ -0,0 +1,223 @@
+storage, along with some extra metadata.
+
+1. Clean up any left over restored entries from previous failed restores.
+2. Gather all data from localForage storage.
+3. Wrap in backup envelope with some metadata about the backup.
+4. Transform data to a format that is safe to serialize to a file.
+5. Transform data in to binary blob data.
+6. Save to file
+
+## File Load Process
+
+This process will convert the binary data in the file to a JSON structure that
+can be restored. The actual restoration process is separate. This step is used
+as a validation of the backup data in the file.
+
+1. Transform binary blob data to file safe JSON data.
+2. Transform file safe JSON data to normal JSON data.
+3. Parse JSON data as backup data.
+
+## Restore Process
+
+Since we don't have transactions, we take a cautious approach to the restore
+process. We attempt to write all entries from the backup file to the localForage
+storage. If this succeeds, then we remove the old entries and rename the
+restored entries.
+
+If there is a failure during the writing of the restore entries then the
+existing database is left in a working state.
+
+1. Clean up any left over restored entries from previous failed restores.
+2. Write all the backup data entries to localForage with "restored" prefix.
+3. Delete all non-restored entries.
+4. Rename all restored entries by removing prefix.
+
+*/
+
+/** Backs up the localDb to a file */
+export async function saveLocalForageToFile(localDb: LocalDb) {
+  const backup = await createBackupData(localDb);
+  const serializedExportData = serializeData(backup);
+  logger.debug("Backup serialized for file", serializedExportData);
+  const fileData = toJsonFileData(serializedExportData);
+  logger.debug(`Converted data to JSON file blob`);
+  const fileName = "graph-explorer-config.json";
+  saveAs(fileData, fileName);
+  logger.debug(`Saved config data as ${fileName}`);
+}
+
+/** Reads the given blob, verifies it is a backup file, restores the data localDb. */
+export async function readBackupDataFromFile(blob: Blob) {
+  const jsonData = await fromFileToJson(blob);
+  logger.debug("Deserialized JSON data from file blob", jsonData);
+  const deserializedData = deserializeData(jsonData);
+  logger.debug(
+    "Deserialized JSON data from JSON serialization",
+    deserializedData
+  );
+  const result = await SerializedBackupSchema.parseAsync(deserializedData);
+  logger.debug("Parsed backup result", result);
+  return result;
+}
+
+/** Applies the given backup data to the localDb, overwriting all contents. */
+export async function restoreBackup(
+  backup: SerializedBackup,
+  localDb: LocalDb
+) {
+  logger.debug("Restoring backup to local db", backup);
+  await removeRestoredEntries(localDb);
+  const entries = applyRestorePrefixToEntries(backup.data);
+  await saveEntriesToDb(entries, localDb);
+  await removeExistingEntries(localDb);
+  await removePrefixFromRestoredEntries(localDb);
```

Comment:
suggestion: Add error handling for each step.

Reply:
The error bubbles up to the UI and is handled there.